File: Data Structures & Algorithms/subsets/submission-2.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        path = []
        ans = []
        n = len(nums)

        def dfs(i):
            if i == n:
                ans.append(path.copy())
                return 
            #选或者不选
            #不选
            dfs(i+1)

            #选
            path.append(nums[i])
            dfs(i+1)
            path.pop()
        dfs(0)
        return ans
            

File: Data Structures & Algorithms/subsets-ii/submission-2.java
class Solution {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        List<Integer> path = new ArrayList<>();
        List<List<Integer>> ans = new ArrayList<>();
        Arrays.sort(nums);
        dfs(0, nums, path, ans);
        return ans;
    }

    private void dfs(int i, int[]nums, List<Integer> path, List<List<Integer>> ans) {
        ans.add(new ArrayList<>(path));
        for(int j = i; j < nums.length; j++) {
            if (j > i && nums[j] == nums[j-1]) continue;
            path.add(nums[j]);
            dfs(j  + 1, nums, path, ans);
            path.remove(path.size()-1);
        }
    }
}
